neqo-transport/src/connection/mod.rs

```diff
@@ -2788,6 +2841,84 @@ impl Connection {
     pub fn stream_keep_alive(&mut self, stream_id: u64, keep: bool) -> Res<()> {
         self.streams.keep_alive(stream_id.into(), keep)
     }
+
+    fn max_dgram_size_configured(&self) -> Res<u64> {
+        let tph = self.tps.borrow();
+        if let Some(r) = &tph.remote {
+            Ok(r.get_integer(tparams::MAX_DATAGRAM_FRAME_SIZE))
+        } else if let Some(r) = &tph.remote_0rtt {
+            Ok(r.get_integer(tparams::MAX_DATAGRAM_FRAME_SIZE))
+        } else {
+            Err(Error::NotAvailable)
+        }
+    }
+
+    #[cfg(test)]
+    pub fn set_max_dgram_size(&mut self, v: u64) {
+        self.tps
+            .borrow_mut()
+            .remote
+            .as_mut()
+            .unwrap()
+            .set_integer(tparams::MAX_DATAGRAM_FRAME_SIZE, v);
+    }
+
+    /// Returns the current max size of a datagram that can fit into a packet.
+    /// # Error
+    /// The function returns `NotAvailable` if datagrams are not enabled.
+    pub fn max_dgram_size(&self) -> Res<u64> {
+        let max_dgram_size = self.max_dgram_size_configured()?;
+        if max_dgram_size == 0 {
+            return Err(Error::NotAvailable);
+        }
+        let version = self.version();
+        let (cspace, tx) = if let Some(crypto) = self
+            .crypto
+            .states
+            .select_tx(PacketNumberSpace::ApplicationData)
+        {
+            crypto
+        } else {
+            return Err(Error::NotAvailable);
+        };
+        let path = self.paths.select_path().ok_or(Error::NotAvailable)?;
+        let mtu = path.borrow().mtu();
+        let encoder = Encoder::with_capacity(mtu);
+
+        let (_, mut builder) = Self::build_packet_header(
+            &path.borrow(),
+            cspace,
+            encoder,
+            tx,
+            &self.address_validation,
+            version,
+            false,
+        );
+        let _ = Self::add_packet_number(
+            &mut builder,
+            tx,
+            self.loss_recovery
+                .largest_acknowledged_pn(PacketNumberSpace::ApplicationData),
+        );
+
+        let data_len_possible =
+            u64::try_from(mtu.saturating_sub(tx.expansion() + builder.len() + 1)).unwrap();
+        Ok(min(data_len_possible, max_dgram_size))
+    }
+
+    /// Returns if there was an unsend datagram that has been dismissed.
+    /// # Error
+    /// The function returns `TooMuchData` if the supply buffer cannot fit into a packet.
+    /// The function returns `NotAvailable` if datagrams are not enabled.
+    pub fn send_dgram(&mut self, buf: &[u8]) -> Res<bool> {
```

Comment:
No point in a contraction here.

```suggestion
    pub fn send_datagram(&mut self, buf: &[u8]) -> Res<bool> {
```

Reply:
Adding an additional parameter for the expiration time of the datagram would be useful, I think.  We might add a new function (`send_datagram_before()`) for that though.

I'd be happy to have that tracked in an issue.

Reply:
Open issue #1203